scripts/copyFilesUtils.mjs

```diff
@@ -91,26 +92,118 @@ export async function typescriptCopy({ from, to }) {
+          },
+          import: {
+            types: `./esm/${baseName}.d.ts`,
+            default: `./esm/${baseName}.js`,
+          },
+          [modernCondition]: {
+            types: `./modern/${baseName}.d.ts`,
+            default: `./modern/${baseName}.js`,
+          },
+          ...rest,
+        },
+      };
+    }
+  }
+
+  if (typeof conditions === 'string' && /\.tsx?$/.test(conditions)) {
+    return createExportFor(exportName, { [srcCondition]: conditions });
+  }
+
+  return {
+    [exportName]: conditions,
+  };
+}
+
+// TODO: remove useEsmExports paramater once X is on the ESM-exports package layout (default to true)
+export async function createPackageFile(useEsmExports = false) {
   const packageData = await fse.readFile(path.resolve(packagePath, './package.json'), 'utf8');
   const { nyc, scripts, devDependencies, workspaces, ...packageDataOther } =
     JSON.parse(packageData);
 
-  const newPackageData = {
-    ...packageDataOther,
-    private: false,
-    ...(packageDataOther.main
-      ? {
-          main: fse.existsSync(path.resolve(buildPath, './node/index.js'))
-            ? './node/index.js'
-            : './index.js',
-          module: fse.existsSync(path.resolve(buildPath, './esm/index.js'))
-            ? './esm/index.js'
-            : './index.js',
-        }
-      : {}),
+  const packageExports = {
+    ...createExportFor('.', { [srcCondition]: './src/index.ts' }),
+    ...createExportFor('./*', { [srcCondition]: './src/*/index.ts' }),
   };
 
+  if (packageDataOther.exports) {
+    for (const [exportName, conditions] of Object.entries(packageDataOther.exports)) {
+      if (conditions) {
+        Object.assign(packageExports, createExportFor(exportName, conditions));
+      } else {
+        delete packageExports[exportName];
+      }
+    }
+  }
+
+  if (polyfillLegacyModern) {
+    const exportedNames = new Set(Object.keys(packageExports));
+    for (const exportedName of exportedNames) {
+      const modernName = exportedName.replace(/^\./, legacyModernPrefix);
+      const modernExport = packageExports[exportedName][modernCondition] ?? null;
+      if (modernExport && !exportedNames.has(modernName)) {
+        packageExports[modernName] = modernExport;
+      }
+    }
+  }
+
+  const newPackageData = useEsmExports
+    ? {
+        ...packageDataOther,
+        private: false,
+        ...(packageDataOther.main
+          ? {
+              main: './index.js',
+              module: './esm/index.js',
+            }
+          : {}),
+        exports: packageExports,
+      }
+    : {
+        ...packageDataOther,
+        private: false,
+        ...(packageDataOther.main
+          ? {
+              main: fse.existsSync(path.resolve(buildPath, './node/index.js'))
+                ? './node/index.js'
+                : './index.js',
+              module: fse.existsSync(path.resolve(buildPath, './esm/index.js'))
+                ? './esm/index.js'
+                : './index.js',
```

Comment:
Can (or should) we standardize the subfolder structure?

Reply:
We do, this is for the X compatibility mode, will be removed after X is on this setup.